src/fetch.rs

```diff
@@ -94,91 +93,71 @@ fn read_latest_version(
-        write!(output, "{:>12}", "Initializing")?;
-        output.reset()?;
-        writeln!(output, " '{}' index", registry)?;
-
-        let mut opts = git2::RepositoryInitOptions::new();
-        opts.bare(true);
-        git2::Repository::init_opts(&registry_path, &opts)?;
-        return Ok(());
-    }
-
-    let repo = git2::Repository::open(&registry_path)?;
+    let index = crates_index::BareIndex::from_url(registry.as_str())?;
+    let mut index = index.open_or_clone()?;
     if !quiet {
         output.set_color(ColorSpec::new().set_fg(Some(Color::Green)).set_bold(true))?;
         write!(output, "{:>12}", "Updating")?;
         output.reset()?;
         writeln!(output, " '{}' index", registry)?;
     }
 
-    let refspec = format!(
-        "refs/heads/{0}:refs/remotes/origin/{0}",
-        get_checkout_name(registry_path)?
-    );
-    fetch_with_cli(&repo, registry.as_str(), &refspec)?;
+    while let Err(err) = index.retrieve() {
+        if need_retry(err)? {
+            registry_blocked_message(&mut output)?;
+            std::thread::sleep(REGISTRY_BACKOFF);
+        } else {
+            break;
+        }
+    }
 
     Ok(())
 }
 
-// https://github.com/rust-lang/cargo/blob/57986eac7157261c33f0123bade7ccd20f15200f/src/cargo/sources/git/utils.rs#L758
-fn fetch_with_cli(repo: &git2::Repository, url: &str, refspec: &str) -> Result<()> {
-    let cmd = subprocess::Exec::shell("git")
-        .arg("fetch")
-        .arg("--tags") // fetch all tags
-        .arg("--force") // handle force pushes
-        .arg("--update-head-ok") // see discussion in rust-lang/cargo#2078
-        .arg(url)
-        .arg(refspec)
-        // If cargo is run by git (for example, the `exec` command in `git
-        // rebase`), the GIT_DIR is set by git and will point to the wrong
-        // location (this takes precedence over the cwd). Make sure this is
-        // unset so git will look at cwd for the repo.
-        .env_remove("GIT_DIR")
-        // The reset of these may not be necessary, but I'm including them
-        // just to be extra paranoid and avoid any issues.
-        .env_remove("GIT_WORK_TREE")
-        .env_remove("GIT_INDEX_FILE")
-        .env_remove("GIT_OBJECT_DIRECTORY")
-        .env_remove("GIT_ALTERNATE_OBJECT_DIRECTORIES")
-        .cwd(repo.path());
-
-    let _ = cmd.capture().map_err(|e| match e {
-        subprocess::PopenError::IoError(io) => ErrorKind::Io(io),
-        _ => unreachable!("expected only io error"),
-    })?;
+const REGISTRY_BACKOFF: Duration = Duration::from_secs(1);
+
+fn need_retry(err: crates_index::Error) -> Result<bool> {
+    match err {
+        crates_index::Error::Git(err) => {
+            if err.class() == git2::ErrorClass::Index && err.code() == git2::ErrorCode::Locked {
+                Ok(true)
+            } else {
+                Err(crates_index::Error::Git(err).into())
+            }
+        }
+        err => Err(err.into()),
+    }
+}
```

Comment:
this function never returns `Ok(false)`, also some docs/comments would be nice
should it take the number of attempts into consideration?

Reply:
The lack of `Ok(false)` was more an API design than.  The other patterns I had thought of for this API (like `Result<()>`) felt like they would be confusing in context.  

I do have one idea to possibly make this a little less weird. I'll give that a try.